src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -453,13 +556,27 @@ DatabasePtr DatabaseCatalog::detachDatabase(ContextPtr local_context, const Stri
     DatabasePtr db;
     {
         std::lock_guard lock{databases_mutex};
-        assertDatabaseExistsUnlocked(database_name);
         db = databases.find(database_name)->second;
-        UUID db_uuid = db->getUUID();
-        if (db_uuid != UUIDHelpers::Nil)
-            removeUUIDMapping(db_uuid);
-        databases.erase(database_name);
     }
+    if (!db)
+    {
+        DatabaseNameHints hints(*this);
+        std::vector<String> names = hints.getHints(database_name);
+        if (names.empty())
+        {
+            throw Exception(ErrorCodes::UNKNOWN_DATABASE, "Database {} does not exist", backQuoteIfNeed(database_name));
+        }
+        else
+        {
+            throw Exception(ErrorCodes::UNKNOWN_DATABASE, "Database {} does not exist. Maybe you meant {}?", backQuoteIfNeed(database_name), backQuoteIfNeed(names[0]));
+        }
+    }
+
+    UUID db_uuid = db->getUUID();
+    if (db_uuid != UUIDHelpers::Nil)
+        removeUUIDMapping(db_uuid);
+    std::lock_guard lock{databases_mutex};
```

Comment:
this operation should be atomic - we should not unlock/lock in the middle - it will allow to creep in some asynchronous operation which can screw up `databases`